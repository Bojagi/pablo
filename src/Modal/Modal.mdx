---
name: Modal
route: /modal
menu: Components
---

import { Props, Playground } from 'docz'
import { Box } from '../Box/Box.tsx'
import { Modal } from './Modal.tsx'
import { Paragraph } from '../Typography/index.ts'
import { Button } from '../Button/index.ts'
import { ButtonBar } from '../ButtonBar/index.ts'
import { IconButton } from '../IconButton/index.ts'
import { X, MoreHorizontal } from 'react-feather';

# Modal

With the `Modal` component you can show additional content on user interaction that overlays the page and
hides the content behind it. The additional content is presented within a pane and can also have additional panes.

## Simple modal

The modal takes the content to display when it's open as a child. The title of the modal is optional.

<Playground>
  {() => {
    const [open, setOpen] = React.useState(false);
    return (
      <>
        <Button onClick={() => setOpen(true)}>open modal</Button>
        <Modal title="Hej" onClose={() => setOpen(false)} open={open}>
          <Paragraph>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
            labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
            laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
            voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
            non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
          </Paragraph>
        </Modal>
      </>
    );
  }}
</Playground>

## Modal as a confirmation dialog

You can use a modal also as confirmation dialog together with the `ButtonBar` component.

<Playground>
  {() => {
    const [open, setOpen] = React.useState(false);
    return (
      <>
        <Button onClick={() => setOpen(true)}>open modal</Button>
        <Modal onClose={() => setOpen(false)} open={open}>
          <Paragraph>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
            labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
            laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
            voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
            non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
          </Paragraph>
          <ButtonBar>
            <Button size="medium" color="plain" variant="primary" onClick={() => setOpen(false)}>
              Cancel
            </Button>
            <Button size="medium" onClick={() => setOpen(false)}>
              Done
            </Button>
          </ButtonBar>
        </Modal>
      </>
    );
  }}
</Playground>

## Top right icon/component

You can put any component in the top right corner of a modal with the `topRightItem` prop (e.g. a close button).

<Playground>
  {() => {
    const [open, setOpen] = React.useState(false);
    return (
      <>
        <Button onClick={() => setOpen(true)}>open modal</Button>
        <Modal
          onClose={() => setOpen(false)}
          title="Hej"
          open={open}
          topRightItem={({ onClose }) => (
            <IconButton size="medium">
              <X onClick={onClose} />
            </IconButton>
          )}
        >
          <Paragraph>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
            labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
            laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
            voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
            non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
          </Paragraph>
          <ButtonBar>
            <Button size="medium" color="plain" variant="primary" onClick={() => setOpen(false)}>
              Cancel
            </Button>
            <Button size="medium" onClick={() => setOpen(false)}>
              Done
            </Button>
          </ButtonBar>
        </Modal>
      </>
    );
  }}
</Playground>

You can also put in multiple items in there (e.g. multiple buttons with a `ButtonBar` component).

<Playground>
  {() => {
    const [open, setOpen] = React.useState(false);
    return (
      <>
        <Button onClick={() => setOpen(true)}>open modal</Button>
        <Modal
          onClose={() => setOpen(false)}
          title="Hej"
          open={open}
          topRightItem={({ onClose }) => (
            <ButtonBar>
              <IconButton size="medium">
                <MoreHorizontal />
              </IconButton>
              <IconButton size="medium" onClick={onClose}>
                <X />
              </IconButton>
            </ButtonBar>
          )}
        >
          <Paragraph>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
            labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
            laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
            voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
            non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
          </Paragraph>
          <ButtonBar>
            <Button size="medium" color="plain" variant="primary" onClick={() => setOpen(false)}>
              Cancel
            </Button>
            <Button size="medium" onClick={() => setOpen(false)}>
              Done
            </Button>
          </ButtonBar>
        </Modal>
      </>
    );
  }}
</Playground>

### Additional panes

Sometimes you need more panes to add a different section to the modal. These additional panes are displayed
as separate boxes below the main box. In case your panes change order or are dynamically added and removed you
need add `key` props to them, otherwise there might be rendering problems.

<Playground>
  {() => {
    const [open, setOpen] = React.useState(false);
    return (
      <>
        <Button onClick={() => setOpen(true)}>open modal</Button>
        <Modal
          onClose={() => setOpen(false)}
          open={open}
          additionalPanes={[
            <Paragraph>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
              labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
              laboris nisi ut aliquip ex ea commodo consequat.
            </Paragraph>,
            <Paragraph mb={0}>Goodbye</Paragraph>,
            <Paragraph mb={0}>I am the Walrus</Paragraph>,
          ]}
        >
          <Paragraph>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
            labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
            laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
            voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
            non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
          </Paragraph>
        </Modal>
      </>
    );
  }}
</Playground>


## Import

```js
import { Modal } from '@bojagi/pablo/Modal';
```

## Props

<Props of={Modal} />
