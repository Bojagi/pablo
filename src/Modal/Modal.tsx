import React from 'react';
import ReactDOM from 'react-dom';
import styled, { css } from 'styled-components';
import { Flex } from '../Box';
import { Title } from '../Typography';
import { getComponentStyle, shadowTransformer, transitionTransformer } from '../utils/styleHelpers';

export interface TopRightItemProps {
  onClose?: () => void;
}

export interface ModalProps {
  children: React.ReactNode;
  title?: React.ReactNode;
  onClose?: () => void;
  additionalPanes?: React.ReactNode[];
  topRightItem?: React.FC<TopRightItemProps>;
  open?: boolean;
}

interface BackdropProps {
  open: boolean;
  onClick: (e: React.PointerEvent<HTMLDivElement>) => void;
}

const Backdrop = styled.div<BackdropProps>`
  position: fixed;
  width: 100vw;
  height: 100vh;
  z-index: 1000;
  left: 0;
  top: 0;
  box-sizing: border-box;
  overflow: scroll;
  background-color: ${getComponentStyle('modal.backdropColor')};
  display: flex;
  justify-content: center;
  align-items: center;
  transition: ${getComponentStyle('modal.backdropTransition', transitionTransformer)};
  ${(props) =>
    props.open
      ? css`
          opacity: 1;
        `
      : css`
          pointer-events: none;
          opacity: 0;
        `}
`;

interface ModalAreaProps {
  open: boolean;
}

const ModalArea = styled.div<ModalAreaProps>`
  ${(props) =>
    !props.open &&
    css`
      transform: ${getComponentStyle('modal.box.closedTransform')};
    `}
  transition: ${getComponentStyle('modal.box.transition', transitionTransformer)};
  max-width: ${getComponentStyle('modal.box.width')};
  margin: auto;
  min-height: min-content;
  padding: ${getComponentStyle('modal.padding')};
`;

const ModalBox = styled.div`
  border-radius: ${getComponentStyle('modal.box.borderRadius')}px;
  background-color: ${getComponentStyle('modal.box.backgroundColor')};
  padding: ${getComponentStyle('modal.box.padding')};
  box-shadow: ${getComponentStyle('modal.box.shadow', shadowTransformer)};
`;

const PaneBox = styled.div`
  margin-top: ${getComponentStyle('modal.gap')};
`;

export function Modal({
  children,
  additionalPanes,
  topRightItem: TopRightItem,
  onClose,
  title,
  open = false,
}: ModalProps) {
  const mountPoint = React.useMemo(() => document.createElement('div'), []);
  React.useEffect(() => {
    mountPoint.setAttribute('data-testid', 'pbl-modal-mountpoint');
    document.body.appendChild(mountPoint);
    return () => {
      document.body.removeChild(mountPoint);
    };
  }, [mountPoint]);

  const mouseDownRef = React.useRef<HTMLElement>(null);
  const handleClose = React.useCallback(
    (e: React.PointerEvent<HTMLDivElement>) => {
      // Only close if the mouse down before the click happened directly over the backdrop
      if (e.target === e.currentTarget && !mouseDownRef.current && onClose) {
        onClose();
      }
      (mouseDownRef.current as any) = null;
    },
    [onClose]
  );

  return ReactDOM.createPortal(
    <Backdrop data-testid="pbl-modal-backdrop" onClick={handleClose} open={open}>
      <ModalArea data-testid="pbl-modal-area" open={open}>
        <ModalBox
          data-testid="pbl-modal-box"
          onMouseDown={(e) => {
            (mouseDownRef.current as any) = e.currentTarget;
          }}
        >
          {(title || TopRightItem) && (
            <Flex data-testid="pbl-modal-title-box" justifyContent="space-between">
              {title && <Title>{title}</Title>}
              {TopRightItem && <TopRightItem onClose={onClose} />}
            </Flex>
          )}
          {children}
        </ModalBox>
        {additionalPanes &&
          additionalPanes.map((pane, i) => (
            <PaneBox key={i} data-testid="pbl-modal-pane">
              <ModalBox data-testid="pbl-modal-pane-box">{pane}</ModalBox>
            </PaneBox>
          ))}
      </ModalArea>
    </Backdrop>,
    mountPoint
  );
}
