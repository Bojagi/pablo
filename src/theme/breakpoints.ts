export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
