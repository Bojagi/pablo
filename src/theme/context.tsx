import * as React from 'react';
import { PabloTheme, ComponentStyles } from './types';

export const pabloThemeContext = React.createContext<PabloTheme>({} as any);
export const pabloComponentStylesContext = React.createContext<ComponentStyles>({} as any);
export const rootContext = React.createContext<Document | ShadowRoot | null>(null);

export const useComponentStyleContext = () => React.useContext(pabloComponentStylesContext);

export const usePabloTheme = () => React.useContext(pabloThemeContext);
