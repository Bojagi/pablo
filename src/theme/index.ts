export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
