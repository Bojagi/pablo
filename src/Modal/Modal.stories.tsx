import React from 'react';

import { MoreHorizontal, X } from 'react-feather';
import { css } from '@emotion/react';
import { Button } from '../Button';
import { ButtonBar } from '../ButtonBar/ButtonBar';
import { IconButton } from '../IconButton';
import { getComponentStyle, getSpacing } from '../styleHelpers';
import { PabloThemeProvider } from '../theme';
import { Paragraph } from '../Typography';
import { Modal } from './Modal';

export default {
  title: 'Modal',
};

const TextContent = () => (
  <Paragraph>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
    laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
    voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
    non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </Paragraph>
);

const ModalButtonBar = ({ setOpen }) => (
  <ButtonBar>
    <Button size="medium" color="plain" variant="primary" onClick={() => setOpen(false)}>
      Cancel
    </Button>
    <Button size="medium" onClick={() => setOpen(false)}>
      Done
    </Button>
  </ButtonBar>
);

const BaseStory = ({ additionalBody: AdditionalBody, ...args }) => {
  const [open, setOpen] = React.useState(false);
  return (
    <>
      <Button onClick={() => setOpen(true)}>open modal</Button>
      <Modal onClose={() => setOpen(false)} open={open} {...args}>
        <TextContent />
        {AdditionalBody && <AdditionalBody setOpen={setOpen} />}
      </Modal>
    </>
  );
};

export const SimpleModal = BaseStory.bind(null);
SimpleModal.args = {
  title: 'Hallo',
};

export const ModalWithButtons = BaseStory.bind(null);
ModalWithButtons.args = {
  title: 'Hallo',
  additionalBody: ModalButtonBar,
};

export const WithTopRightItem = BaseStory.bind(null);
WithTopRightItem.args = {
  title: 'Hallo',
  topRightItem: ({ onClose }) => (
    <IconButton size="medium">
      <X onClick={onClose} />
    </IconButton>
  ),
  additionalBody: ModalButtonBar,
};

export const WithMultipleTopRightItems = BaseStory.bind(null);
WithMultipleTopRightItems.args = {
  title: 'Hallo',
  topRightItem: ({ onClose }) => (
    <ButtonBar>
      <IconButton size="medium">
        <MoreHorizontal />
      </IconButton>
      <IconButton size="medium" onClick={onClose}>
        <X />
      </IconButton>
    </ButtonBar>
  ),
  additionalBody: ModalButtonBar,
};

export const WithoutTitle = BaseStory.bind(null);
WithoutTitle.args = {
  additionalBody: ModalButtonBar,
};

export const WithAdditionalPanes = BaseStory.bind(null);
WithAdditionalPanes.args = {
  title: 'Hallo',
  topRightItem: ({ onClose }) => (
    <IconButton size="medium">
      <X onClick={onClose} />
    </IconButton>
  ),
  additionalPanes: [
    <>
      <TextContent />
      <TextContent />
      <TextContent />
    </>,
    <Paragraph mb={0}>Goodbye</Paragraph>,
    <Paragraph mb={0}>I am the Walrus</Paragraph>,
  ],
  additionalBody: ModalButtonBar,
};

export const WithCustomStyles = BaseStory.bind(null);
WithCustomStyles.args = {
  title: 'Hallo',
  topRightItem: ({ onClose }) => (
    <IconButton size="medium">
      <X onClick={onClose} />
    </IconButton>
  ),
  additionalPanes: [
    <>
      <TextContent />
      <TextContent />
      <TextContent />
    </>,
    <Paragraph mb={0}>Goodbye</Paragraph>,
    <Paragraph mb={0}>I am the Walrus</Paragraph>,
  ],
  additionalBody: ModalButtonBar,
  customStyles: {
    backdrop: css`
      background-color: rgba(0, 0, 255, 0.2);
    `,
    box: css`
      border: 5px solid red;
    `,
    area: (props) => css`
      background-color: rgba(0, 0, 255, 0.2);
      border-radius: ${getComponentStyle('modal.box.borderRadius')(props)}px;
    `,
    paneBox: (props) => css`
      border: 5px solid blue;
      border-radius: ${getComponentStyle('modal.box.borderRadius')(props)}px;
      padding: ${getSpacing(4)(props)};
    `,
  },
};

export const WithCustomStylesFromTheme = () => {
  const [open, setOpen] = React.useState(false);
  const args = {
    title: 'Hallo',
    topRightItem: ({ onClose }) => (
      <IconButton size="medium">
        <X onClick={onClose} />
      </IconButton>
    ),
    additionalPanes: [
      <>
        <TextContent />
        <TextContent />
        <TextContent />
      </>,
      <Paragraph mb={0}>Goodbye</Paragraph>,
      <Paragraph mb={0}>I am the Walrus</Paragraph>,
    ],
    customStyles: {
      backdrop: css`
        background-color: rgba(0, 0, 255, 0.2);
      `,
      box: css`
        border: 5px solid red;
      `,
      area: (props) => css`
        background-color: rgba(0, 0, 255, 0.2);
        border-radius: ${getComponentStyle('modal.box.borderRadius')(props)}px;
      `,
      paneBox: (props) => css`
        border: 5px solid blue;
        border-radius: ${getComponentStyle('modal.box.borderRadius')(props)}px;
        padding: ${getSpacing(4)(props)};
      `,
    },
  };

  return (
    <PabloThemeProvider
      componentStyles={{
        modal: {
          styles: {
            box: css`
              padding: 200px;
            `,
          },
        },
      }}
    >
      <Button onClick={() => setOpen(true)}>open modal</Button>
      <Modal onClose={() => setOpen(false)} open={open} {...args}>
        <TextContent />
      </Modal>
    </PabloThemeProvider>
  );
};
