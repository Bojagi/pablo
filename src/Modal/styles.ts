import { getSpacing } from '../utils/styleHelpers/getSpacing';
import { Style } from '../theme/types';

export interface ModalStyles {
  backdropColor: Style;
  padding: Style;
  gap: Style;
  box: {
    borderRadius: number;
    width: Style;
    backgroundColor: Style;
    padding: Style;
  };
}

export const modalStyles: ModalStyles = {
  backdropColor: 'rgba(5, 0, 21, 0.5)',
  padding: getSpacing(3),
  gap: getSpacing(1.5),
  box: {
    backgroundColor: '#fafafa',
    padding: getSpacing(1.5),
    width: '780px',
    borderRadius: 12,
  },
};
