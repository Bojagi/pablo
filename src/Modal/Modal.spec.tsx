import {
  render,
  getByTestId as globalGetByTestId,
  getAllByTestId as globalGetAllByTestId,
  fireEvent,
  act,
} from '@testing-library/react';
import React from 'react';
import { PabloThemeProvider } from '../theme';
import { Modal } from './Modal';

test('Render closed modal with just a content element', () => {
  const { container, getByTestId } = renderComponent({ children: <div>Hello World</div> });
  expect(container).toMatchSnapshot();

  // Backdrop
  expect(getByTestId('pbl-modal-backdrop')).toHaveStyleRule('width', '100vw');
  expect(getByTestId('pbl-modal-backdrop')).toHaveStyleRule('height', '100vh');
  expect(getByTestId('pbl-modal-backdrop')).toHaveStyleRule('background-color', 'rgba(0,0,0,0.6)');
  expect(getByTestId('pbl-modal-backdrop')).toHaveStyleRule('display', 'flex');
  expect(getByTestId('pbl-modal-backdrop')).toHaveStyleRule('justify-content', 'center');
  expect(getByTestId('pbl-modal-backdrop')).toHaveStyleRule('align-items', 'center');
  expect(getByTestId('pbl-modal-backdrop')).toHaveStyleRule('opacity', '0');
  // Click though
  expect(getByTestId('pbl-modal-backdrop')).toHaveStyleRule('pointer-events', 'none');

  // Modal Area
  expect(getByTestId('pbl-modal-area')).toHaveStyleRule('max-width', '780px');
  expect(getByTestId('pbl-modal-area')).toHaveStyleRule('padding', '24px');
  expect(getByTestId('pbl-modal-area')).toHaveStyleRule('margin', 'auto'); // Crucial for centering and scrolling
  expect(getByTestId('pbl-modal-area')).toHaveStyleRule('min-height', 'min-content'); // Crucial for centering and scrolling

  // ModalBox
  expect(getByTestId('pbl-modal-box')).toHaveStyleRule('border-radius', '12px');
  expect(getByTestId('pbl-modal-box')).toHaveStyleRule('background-color', '#fafafa');
  expect(getByTestId('pbl-modal-box')).toHaveStyleRule('padding', '12px');
});

test('Render open modal', () => {
  const { getByTestId } = renderComponent({ open: true, children: <div>Hello World</div> });
  expect(getByTestId('pbl-modal-backdrop')).toHaveStyleRule('opacity', '1');
  // Not click though
  expect(getByTestId('pbl-modal-backdrop')).not.toHaveStyleRule('pointer-events');
});

test('Render with title', () => {
  const { getByTestId } = renderComponent({
    open: true,
    title: (
      <>
        Hans <strong>Wurst</strong>
      </>
    ),
    children: <div>Hello World</div>,
  });
  expect(getByTestId('pbl-modal-box')).toMatchSnapshot();
  expect(getByTestId('pbl-modal-title-box')).toHaveTextContent('Hans Wurst');
});

test('Render with topRightItem', () => {
  const { getByTestId } = renderComponent({
    open: true,
    topRightItem: () => <div>something</div>,
    children: <div>Hello World</div>,
  });
  expect(getByTestId('pbl-modal-box')).toMatchSnapshot();
  expect(getByTestId('pbl-modal-title-box')).toHaveTextContent('something');
});

test('Render with title and topRightItem', () => {
  const { getByTestId } = renderComponent({
    open: true,
    title: (
      <>
        Hans <strong>Wurst</strong>
      </>
    ),
    topRightItem: () => <div>something</div>,
    children: <div>Hello World</div>,
  });
  expect(getByTestId('pbl-modal-box')).toMatchSnapshot();
  expect(getByTestId('pbl-modal-title-box')).toHaveTextContent('Hans Wurstsomething');
});

test('Render with additional panes', () => {
  const { getByTestId, getAllByTestId } = renderComponent({
    open: true,
    children: <div>Hello World</div>,
    additionalPanes: [<div>Pane 1</div>, <div>Pane 2</div>, <div>Pane 3</div>],
  });
  expect(getByTestId('pbl-modal-area')).toMatchSnapshot();
  expect(getAllByTestId('pbl-modal-pane')).toBeArrayOfSize(3);
  expect(getAllByTestId('pbl-modal-pane')[0]).toHaveStyleRule('margin-top', '12px');
});

test('Close when clicking on the backdrop', () => {
  const onCloseMock = jest.fn();
  const { getByTestId } = renderComponent({
    open: true,
    onClose: onCloseMock,
    children: <div>Hello World</div>,
  });

  expect(onCloseMock).toHaveBeenCalledTimes(0);
  act(() => {
    fireEvent.click(getByTestId('pbl-modal-backdrop'));
  });
  expect(onCloseMock).toHaveBeenCalledTimes(1);
});

test('Do not close when pressing mouse button on box, moving to backdrop and releasing button there', () => {
  const onCloseMock = jest.fn();
  const { getByTestId } = renderComponent({
    open: true,
    onClose: onCloseMock,
    children: <div>Hello World</div>,
  });

  expect(onCloseMock).toHaveBeenCalledTimes(0);
  // Starting click on box
  act(() => {
    fireEvent.mouseDown(getByTestId('pbl-modal-box'));
  });
  // Releasing click over modal should not trigger click event
  act(() => {
    fireEvent.click(getByTestId('pbl-modal-backdrop'));
  });
  expect(onCloseMock).toHaveBeenCalledTimes(0);
});

test('Click on box should not trigger close event', () => {
  const onCloseMock = jest.fn();
  const { getByTestId } = renderComponent({
    open: true,
    onClose: onCloseMock,
    children: <div>Hello World</div>,
  });

  expect(onCloseMock).toHaveBeenCalledTimes(0);
  act(() => {
    fireEvent.click(getByTestId('pbl-modal-box'));
  });
  expect(onCloseMock).toHaveBeenCalledTimes(0);
});

test('Remove mount point on unmount', () => {
  const { container, unmount } = renderComponent({
    open: true,
    children: <div>Hello World</div>,
  });

  expect(container.querySelector('div[data-testid="pbl-modal-mountpoint"')).not.toBeNull();
  unmount();
  expect(container).toMatchSnapshot();
  expect(container.querySelector('div[data-testid="pbl-modal-mountpoint"')).toBeNull();
});

function renderComponent(props) {
  const { baseElement, unmount } = render(
    <PabloThemeProvider>
      <Modal {...props} />
    </PabloThemeProvider>
  );

  return {
    getByTestId: globalGetByTestId.bind(null, baseElement),
    getAllByTestId: globalGetAllByTestId.bind(null, baseElement),
    unmount,
    container: baseElement,
  };
}
