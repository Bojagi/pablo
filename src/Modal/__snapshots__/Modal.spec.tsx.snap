// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remove mount point on unmount 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
</body>
`;

exports[`Render closed modal with just a content element 1`] = `
.c0 {
  box-sizing: border-box;
  position: fixed;
  width: 100vw;
  height: 100vh;
  z-index: 1000;
  left: 0;
  top: 0;
  overflow: scroll;
  background-color: rgba(0,0,0,0.6);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  pointer-events: none;
  opacity: 0;
}

.c1 {
  box-sizing: border-box;
  -webkit-transform: translateY(60px);
  -ms-transform: translateY(60px);
  transform: translateY(60px);
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
  max-width: 780px;
  margin: auto;
  min-height: -webkit-min-content;
  min-height: -moz-min-content;
  min-height: min-content;
  padding: 24px;
}

.c2 {
  box-sizing: border-box;
  border-radius: 12px;
  background-color: #fafafa;
  padding: 12px;
  box-shadow: 0px 4px 20px rgba(0,0,0,0.2);
}

<body
  style="overflow: auto;"
>
  <div />
  <div
    data-testid="pbl-modal-mountpoint"
  >
    <div
      class="c0"
      data-testid="pbl-modal-backdrop"
    >
      <div
        class="c1"
        data-testid="pbl-modal-area"
      >
        <div
          class="c2"
          data-testid="pbl-modal-box"
        >
          <div>
            Hello World
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Render with additional panes 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
  max-width: 780px;
  margin: auto;
  min-height: -webkit-min-content;
  min-height: -moz-min-content;
  min-height: min-content;
  padding: 24px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 12px;
  background-color: #fafafa;
  padding: 12px;
  box-shadow: 0px 4px 20px rgba(0,0,0,0.2);
}

.c2 {
  box-sizing: border-box;
  margin-top: 12px;
}

<div
  class="c0"
  data-testid="pbl-modal-area"
  open=""
>
  <div
    class="c1"
    data-testid="pbl-modal-box"
  >
    <div>
      Hello World
    </div>
  </div>
  <div
    class="c2"
    data-testid="pbl-modal-pane"
  >
    <div
      class="c1"
      data-testid="pbl-modal-pane-box"
    >
      <div>
        Pane 1
      </div>
    </div>
  </div>
  <div
    class="c2"
    data-testid="pbl-modal-pane"
  >
    <div
      class="c1"
      data-testid="pbl-modal-pane-box"
    >
      <div>
        Pane 2
      </div>
    </div>
  </div>
  <div
    class="c2"
    data-testid="pbl-modal-pane"
  >
    <div
      class="c1"
      data-testid="pbl-modal-pane-box"
    >
      <div>
        Pane 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Render with title 1`] = `
.c1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  box-sizing: border-box;
  font-family: "IBM Plex Sans",sans-serif;
  font-style: normal;
  font-weight: normal;
  margin: 0;
  font-size: 1.5rem;
  line-height: 1.3333333em;
  margin-bottom: 0.5em;
}

.c0 {
  box-sizing: border-box;
  border-radius: 12px;
  background-color: #fafafa;
  padding: 12px;
  box-shadow: 0px 4px 20px rgba(0,0,0,0.2);
}

<div
  class="c0"
  data-testid="pbl-modal-box"
>
  <div
    class="c1"
    data-testid="pbl-modal-title-box"
  >
    <h3
      class="c2"
    >
      Hans 
      <strong>
        Wurst
      </strong>
    </h3>
  </div>
  <div>
    Hello World
  </div>
</div>
`;

exports[`Render with title and topRightItem 1`] = `
.c1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  box-sizing: border-box;
  font-family: "IBM Plex Sans",sans-serif;
  font-style: normal;
  font-weight: normal;
  margin: 0;
  font-size: 1.5rem;
  line-height: 1.3333333em;
  margin-bottom: 0.5em;
}

.c0 {
  box-sizing: border-box;
  border-radius: 12px;
  background-color: #fafafa;
  padding: 12px;
  box-shadow: 0px 4px 20px rgba(0,0,0,0.2);
}

<div
  class="c0"
  data-testid="pbl-modal-box"
>
  <div
    class="c1"
    data-testid="pbl-modal-title-box"
  >
    <h3
      class="c2"
    >
      Hans 
      <strong>
        Wurst
      </strong>
    </h3>
    <div>
      something
    </div>
  </div>
  <div>
    Hello World
  </div>
</div>
`;

exports[`Render with topRightItem 1`] = `
.c1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  box-sizing: border-box;
  border-radius: 12px;
  background-color: #fafafa;
  padding: 12px;
  box-shadow: 0px 4px 20px rgba(0,0,0,0.2);
}

<div
  class="c0"
  data-testid="pbl-modal-box"
>
  <div
    class="c1"
    data-testid="pbl-modal-title-box"
  >
    <div>
      something
    </div>
  </div>
  <div>
    Hello World
  </div>
</div>
`;
